ampd/src/health_check.rs

```diff
@@ -0,0 +1,99 @@
+use error_stack::{Result, ResultExt};
+use std::net::SocketAddr;
+use thiserror::Error;
+
+use axum::{http::StatusCode, routing::get, Json, Router};
+use serde::{Deserialize, Serialize};
+use tokio_util::sync::CancellationToken;
+
+pub struct Server {
+    listener: tokio::net::TcpListener,
+}
+
+#[derive(Error, Debug)]
+pub enum HealthCheckError {
+    #[error("Health check server error: {0}")]
+    Error(String),
+}
+
+impl Server {
+    pub async fn new(bind_addr: SocketAddr) -> Result<Self, HealthCheckError> {
+        Ok(Self {
+            listener: tokio::net::TcpListener::bind(bind_addr)
+                .await
+                .change_context(HealthCheckError::Error(format!(
+                    "Failed binding to addr: {}",
+                    bind_addr
+                )))?,
+        })
+    }
+
+    #[allow(dead_code)]
+    pub fn listening_addr(&self) -> Result<SocketAddr, HealthCheckError> {
+        Ok(self
+            .listener
+            .local_addr()
+            .map_err(|e| HealthCheckError::Error(e.to_string()))?)
+    }
+
+    pub async fn run(self, cancel: CancellationToken) -> Result<(), HealthCheckError> {
+        tracing_subscriber::fmt::init();
```

Comment:
The `main.rs` file already initialises the tracing subscriber. afaik initialising it the second time will result in a panic

Reply:
Thanks ! i was missing to cleanup that part. Fixed tracing at 9937cd5efe25c104a8c9e24b4cd46429b69d6c40